airflow/utils/log/bluecore_bq_gcs_task_handler.py

```diff
@@ -0,0 +1,194 @@
+
+
+
+class BQGCSTaskHandler(GCSTaskHandler):
+    """
+    BQGCSTaskHandler is a python log handler that handles and reads
+    task instance logs. It extends airflow FileTaskHandler and GCSTaskHandler
+    uploads to and reads from GCS remote storage. Upon log reading
+    failure, it reads from host machine's local disk. The logs will be displayed in
+    the airflow UI once retrieved from BigQuery.
+    """
+
+    def __init__(self, base_log_folder, gcs_log_folder, filename_template):
+        super(BQGCSTaskHandler, self).__init__(base_log_folder, gcs_log_folder, filename_template)
+
+    def _read(self, ti, try_number):
+        """
+        Read logs of given task instance and try_number from GCS.
+        If failed, read the log from BigQuery.
+
+        :param ti: task instance object
+        :param try_number: task instance try_number to read logs from
+        """
+        # Explicitly getting log relative path is necessary as the given
+        # task instance might be different than task instance passed in
+        # in set_context method.
+        log_relative_path = self._render_filename(ti, try_number + 1)
+        remote_loc = os.path.join(self.remote_base, log_relative_path)
+
+        if self.gcs_log_exists(remote_loc):
+            # If GCS remote file exists, we do not fetch logs from task instance
+            # local machine even if there are errors reading remote logs, as
+            # remote_log will contain error message.
+            remote_log = self.gcs_read(remote_loc, return_error=True)
+            log = '*** Reading remote log from {}.\n{}\n'.format(
+                remote_loc, remote_log)
+            return log
+
+        # everything above is copied from gcs_task_handler.py, the following elif statement includes changes
+        # specifically for bluecore_gcs_task_handler
+        elif ti.operator == "KubernetesJobOperator":
+            # If remote file is not available and the job is a Kubernetes Job,
+            # use the task instance attributes to query BigQuery logs and
+            # make the logs available in real time from the "Log" page of the UI.
+
+            # TODO:  job_name is always none on the first try, should fix this so an exception doesn't show each time
+            job_name = ti.xcom_pull(task_ids=ti.task_id, key='kubernetes_job_name')
+            if job_name is None:
+                raise Exception("There is no XCOM kubernetes_job_name. Something is exploding.")
+
+            pod_output = self.get_pods(job_name)
+
+            billing_project_id = configuration.get('core', 'kubernetes_bigquery_billing_project')
+            client = bigquery.Client(billing_project_id)
+
+            # get info for the BQ table name
+            data_project_name = configuration.get('core', 'kubernetes_bigquery_data_project')
+            # get BQ table's date (today).. based on insert time & not execution time.
+            bq_table_date = datetime.utcnow().date().strftime("%Y%m%d")
```

Comment:
This is a really good point... since it logs everything from one query, I need to think of a way to account for this.

Reply:
I think this would be an excellent use of [BigQuery table suffixes](https://cloud.google.com/bigquery/docs/querying-wildcard-tables#scanning_a_range_of_tables_using_table_suffix). You can limit by suffixes between the date of the start time of the task (maybe minus 10 minutes for safety) and the current date.